src/i2c.rs

```diff
@@ -0,0 +1,369 @@
+//! Inter-Integrated Circuit (I2C) bus
+
+use afio::MAPR;
+use gpio::{Alternate, OpenDrain};
+use gpio::gpiob::{PB10, PB11, PB6, PB7, PB8, PB9};
+use hal::blocking::i2c::{Read, Write, WriteRead};
+use rcc::{APB1, Clocks};
+use stm32f103xx::{I2C1, I2C2};
+
+/// I2C error
+#[derive(Debug)]
+pub enum Error {
+    /// Timed out busy waiting
+    Timeout,
+    /// Bus error
+    Bus,
+    /// Arbitration loss
+    Arbitration,
+    /// No ack received
+    Acknowledge,
+    /// Overrun/underrun
+    Overrun,
+    // Pec, // SMBUS mode only
+    // Timeout, // SMBUS mode only
+    // Alert, // SMBUS mode only
+    #[doc(hidden)] _Extensible,
+}
+
+#[derive(Debug)]
+pub enum DutyCycle {
+    Ratio1to1,
+    Ratio16to9,
+}
+
+#[derive(Debug)]
+pub enum Mode {
+    Standard { frequency: u32 },
+    Fast { frequency: u32, duty_cycle: DutyCycle },
+}
+
+impl Mode {
+    pub fn get_frequency(&self) -> u32 {
+        match self {
+            &Mode::Standard { frequency } => frequency,
+            &Mode::Fast { frequency, .. } => frequency,
+        }
+    }
+}
+
+
+pub trait Pins<I2C> {
+    const REMAP: bool;
+}
+
+impl Pins<I2C1>
+for (
+    PB6<Alternate<OpenDrain>>,
+    PB7<Alternate<OpenDrain>>,
+) {
+    const REMAP: bool = false;
+}
+
+impl Pins<I2C1>
+for (
+    PB8<Alternate<OpenDrain>>,
+    PB9<Alternate<OpenDrain>>,
+) {
+    const REMAP: bool = true;
+}
+
+impl Pins<I2C2>
+for (
+    PB10<Alternate<OpenDrain>>,
+    PB11<Alternate<OpenDrain>>,
+) {
+    const REMAP: bool = false;
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c<I2C, PINS> {
+    i2c: I2C,
+    pins: PINS,
+    bus_timeout: u32,
+    mode: Mode,
+    pclk1: u32,
+}
+
+impl<PINS> I2c<I2C1, PINS> {
+    pub fn i2c1(
+        i2c: I2C1,
+        pins: PINS,
+        mapr: &mut MAPR,
+        mode: Mode,
+        clocks: Clocks,
+        apb: &mut APB1,
+    ) -> Self
+        where
+            PINS: Pins<I2C1>,
+    {
+        mapr.mapr().modify(|_, w| w.i2c1_remap().bit(PINS::REMAP));
+        I2c::_i2c1(i2c, pins, mode, clocks, apb)
+    }
+}
+
+impl<PINS> I2c<I2C2, PINS> {
+    pub fn i2c2(
+        i2c: I2C2,
+        pins: PINS,
+        mode: Mode,
+        clocks: Clocks,
+        apb: &mut APB1,
+    ) -> Self
+        where
+            PINS: Pins<I2C2>,
+    {
+        I2c::_i2c2(i2c, pins, mode, clocks, apb)
+    }
+}
+
+
+macro_rules! busy_wait {
+    ($cycles:expr, $i2c:expr, $flag:ident) => {
+        {
+            let mut cycles_left = $cycles;
+            loop {
+                if cycles_left == 0 {
```

Comment:
I'm not 100% sold on the idea of having timeout behavior by default. It seems to not be configurable and not possible to opt out of it. Does the I2C spec say anything about timeouts? iirc, SMBus had some timeout feature but I'm not sure if I2C did.

IMO, timeout should be achieved via composition in a generic fashion: you grab something that returns a generator and an `impl Countdown` and you turn that into a blocking operation with timeout. Kind of like the second example on [these docs](https://docs.rs/embedded-hal/0.1.2/embedded_hal/#generic-programming-and-higher-level-abstractions) but using generators.

Reply:
So, an option here would be to provide a generators based API and an implementation of the blocking::i2c traits on top of that. Then the user would be able to use the generic function I mentioned above to get timeout behavior.

Another option would be adding type state to I2c *or* adding a two field `structWithTimeout { i2c: I2c, timer: impl CountDown }`. This second type would implement the blocking::i2c traits and provide timeout behavior; that way you would be able to use that type with generic drivers. `I2c` itself would also implement the blocking::i2c traits; that implementation wouldn't have timeout behavior.

Reply:
The timeout here is critical to normal operation of I2C bus. Without it almost any electrical issue on the bus will hang the micro.

There's only so much I can do without implementing non-blocking version, which I'm gonna try this week.